src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HighCPUShardRca.java

```diff
@@ -0,0 +1,209 @@
+        counter += 1;
+
+        // Populate the Resource HashMaps
+        consumeMetrics(cpuUsageMetrics, cpuUsageHashMap);
+        consumeMetrics(ioTotThroughputMetrics, ioTotThroughputHashMap);
+        consumeMetrics(ioTotSyscallRateMetrics, ioTotSyscallRateHashMap);
+
+        if (counter == rcaPeriod) {
+            ResourceContext context = new ResourceContext(Resources.State.HEALTHY);
+            List<HotShardSummary> HotShardSummaryList = new ArrayList<>();
+            ClusterDetailsEventProcessor.NodeDetails currentNode = ClusterDetailsEventProcessor.getCurrentNodeDetails();
+
+            Set<IndexShardKey> indexShardKeySet = new HashSet<>(cpuUsageHashMap.keySet());
+            indexShardKeySet.addAll(ioTotThroughputHashMap.keySet());
+            indexShardKeySet.addAll(ioTotSyscallRateHashMap.keySet());
+
+            for (IndexShardKey indexShardKey : indexShardKeySet) {
+                double avgCpuUsage = fetchUsageValueFromMap(cpuUsageHashMap, indexShardKey);
+                double avgIoTotThroughput = fetchUsageValueFromMap(ioTotThroughputHashMap, indexShardKey);
+                double avgIoTotSyscallRate = fetchUsageValueFromMap(ioTotSyscallRateHashMap, indexShardKey);
+
+                if (avgCpuUsage > CPU_USAGE_THRESHOLD
+                        || avgIoTotThroughput > IO_TOT_THROUGHPUT_THRESHOLD_IN_BYTES
+                        || avgIoTotSyscallRate > IO_TOT_SYSCALL_RATE_THRESHOLD_PER_SECOND) {
+                    HotShardSummaryList.add(new HotShardSummary(
+                            indexShardKey.getIndexName(), indexShardKey.getShardId(), currentNode.getId(),
+                            avgCpuUsage, CPU_USAGE_THRESHOLD, avgIoTotThroughput, IO_TOT_THROUGHPUT_THRESHOLD_IN_BYTES,
+                            avgIoTotSyscallRate, IO_TOT_SYSCALL_RATE_THRESHOLD_PER_SECOND, SLIDING_WINDOW_IN_SECONDS));
+                    context = new ResourceContext(Resources.State.UNHEALTHY);
+                    LOG.debug("Hot Shard Identified, Shard : {} , avgCpuUsage = {} , avgIoTotThroughput = {}, "
+                            + "avgIoTotSyscallRate = {}", indexShardKey, avgCpuUsage, avgIoTotThroughput, avgIoTotSyscallRate);
+                }
+            }
+
+            // reset the variables
+            counter = 0;
+
+            HotNodeSummary summary = new HotNodeSummary(
+                    currentNode.getId(), currentNode.getHostAddress(), HotShardSummaryList);
+
+            LOG.debug("High CPU Usage Shard RCA Context :  " + context.toString());
+            return new ResourceFlowUnit(this.clock.millis(), context, summary);
+        } else {
+            LOG.debug("Empty FlowUnit returned for High CPU Usage Shard RCA");
+            return new ResourceFlowUnit(this.clock.millis());
+        }
+    }
+
+  @Override
+  public void generateFlowUnitListFromWire(FlowUnitOperationArgWrapper args) {
+    LOG.error("rca: {} should be executed from Local", this.getClass().getSimpleName());
```

Comment:
It might be preferable to throw illegalState exception and capture the stack trace ?

Reply:
Agreed. makes more sense